Typescript migration

hegedus-mark opened the pull request.

hegedus-mark committed:
chore: Add typescript support, eslint, prettier plugin, nvmrc

hegedus-mark committed:
chore: move files to src, and convert index.js to index.ts

hegedus-mark committed:
chore: rewrite command files to ts

hegedus-mark committed:
chore: add deploy-commands script and change deploy-commands to ts

hollowM1ke commented on src/deploy-commands.ts:
```diff
@@ -23,7 +23,7 @@ for (const folder of commandFolders) {
 	for (const file of commandFiles) {
 		const filePath = path.join(commandsPath, file);
-		const { command } = await import(pathToFileURL(filePath));
+		const { command } = await import(pathToFileURL(filePath).toString());
```
why did you use toString()?

A-Cer23 replied:
Using the `import()` method requires the input to be a string. Removing the `.toString()` will cause an error for the `import()` method. This is fine

hollowM1ke commented on src/commands/utility/ping.ts:
```diff
@@ -1,10 +1,10 @@
-import { SlashCommandBuilder } from "discord.js";
+import {ChatInputCommandInteraction, CommandInteraction, SlashCommandBuilder} from "discord.js";
```
I think we only need to import the SlashCommandBuilder and ChatInputCommandInteraction.

A-Cer23 replied:
remove `CommandInteraction` import as it is not being used in the file

hegedus-mark replied:
fixed

hollowM1ke commented on src/commands/utility/server.ts:
```diff
@@ -1,11 +1,11 @@
-import { SlashCommandBuilder } from 'discord.js';
+import {ChatInputCommandInteraction, CommandInteraction, SlashCommandBuilder} from 'discord.js';
```
same here

A-Cer23 replied:
remove `CommandInteraction` import as it is not being used in the file

hegedus-mark replied:
fixed

hollowM1ke commented on src/index.ts:
```diff
@@ -1,13 +1,20 @@
 import { fileURLToPath, pathToFileURL } from 'url';
 
-// eslint-disable-next-line no-undef
+class ClientWithCommands extends Client {
```
Can't we just declare ```commands```  like this:

```const commands: Collection<string, unknown> = new Collection(); ```

And then we would use it as usual?

A-Cer23 replied:
It is recommended by discord.js to extend the client class when using typescript. This is okay.

hollowM1ke reviewed (commented):
In general I think it looks good. I've added some comments I think you should look at.

A-Cer23 commented on .gitignore:
missing `dist` directory

hegedus-mark replied:
fixed

A-Cer23 commented on src/commands/utility/server.ts:
```diff
@@ -1,11 +1,11 @@
-    async execute (interaction) {
-        await interaction.reply(`This server is ${interaction.guild.name} and has ${interaction.guild.memberCount} members.`);
+    async execute (interaction: ChatInputCommandInteraction) {
+        await interaction.reply(`This server is ${interaction.guild?.name} and has ${interaction.guild?.memberCount} members.`);
```
Instead of relying on `?`, do an actual undefined check. Current method will lead the bot to respond with `undefined` for both variables if the bot doesn't have proper permissions on the server

hegedus-mark replied:
I doubt it is a permission issue, I think this could be an issue when the bot runs outside a guild context, no? 
I set the context and added a condition

A-Cer23 commented on src/index.ts:
```diff
@@ -38,14 +45,16 @@ client.once(Events.ClientReady, readyClient => {
   }
 
   try {
+    // @ts-ignore
```
This is okay as a temporary solution.

An example fix, emphasis on just being an example.

```
type Command = {
  execute(interaction: Interaction): void
}

class ClientWithCommands extends Client {
  commands: Collection<string, Command>

  constructor() {
    super({intents: [GatewayIntentBits.Guilds]});
    this.commands = new Collection();
  }
}
```

hegedus-mark replied:
It was indeed a temporary solution, but I did add your suggestion for now. We'll eventually have to restructure the index.ts

A-Cer23 reviewed (changes requested):
`dist` directory to be included in `.gitignore`
unused imports to be removed
change required on how to handle possible undefined values in `src/commands/utility/server.ts`

hegedus-mark committed:
chore: Add dist to gitignore

hegedus-mark committed:
chore: remove unused imports

hegedus-mark committed:
fix: Add Command type, this is just a temporary solution

hegedus-mark committed:
fix: Change server.ts to only run in Guild context

Note: I assume interaction.guild can be null, when the bot runs outside a guild context

hegedus-mark commented:
I made the requested changes and made 2 issues #14 #15, that would probably solve some headaches in the future

hegedus-mark committed:
fix: replace previous eslint config with working version

The previous eslint config wasn't configured properly for flat config, so I had to fix that.

hegedus-mark committed:
chore: Add missing Guild_ID to Sample.env

hegedus-mark committed:
fix: Replace member with name
